PR to add code signing to the ci workflow

sconklin opened the pull request:
This PR adds code signing for Windows executables using a private key hosted on Google HSM.

The new workflow signs seamly2d.exe, seamlyme.exe, and the installer exe file.

According to best practices to avoid signing of malicious code, signing requires an approval step. Seamly core developers have the ability to approve.

In case of an emergency requiring release of unsigned executables, there is an environment variable that can be set to bypass signing. This might be used (for example) if google HSM is down, or our cert expires and we still want to release unsigned images.

sconklin committed:
in progress plan for workflow code signing

sconklin committed:
revised implementation plan for code signing workflow

sconklin committed:
Add files needed for signing

sconklin committed:
signing script now tested, added initial signing workflow yaml file

sconklin committed:
Windows Code Signing workflow ready for testing

sconklin committed:
Remove local unsigned test executable

sconklin committed:
complete signing workflow for test branch

Added build and signing workflow for TESTING_GUIDE
Added approval steps for code signing
Added a way to bypass signing

sconklin committed:
change a file to force a test build

sconklin committed:
Remove comment

sconklin committed:
Change ci workflow to add signing

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/FashionFreedom/Seamly2D?pullRequest=1327) <br/>All committers have signed the CLA.

sconklin commented:
I'm definitely going to remove some of the markdown files which don't contribute

csett86 reviewed (commented):
I have some questions on this PR, unfortunately:

Who controls the Google HSM that is used here for signing, ie who controls the keys used to sign?

Why is Google HSM used in the first place?

Specifically this PR then also adds a lot of workflows like "simple-test", "simple-test-2", manual-test etc which all seem like testing leftovers and only the adaption to the main "ci.yml" workflow should remain.

In addition, as you said @sconklin, the readmes seem quite verbose...

csett86 commented:
Overall I would have expected a lighter setup like Azure Trusted Signing (https://learn.microsoft.com/en-us/azure/trusted-signing/quickstart?tabs=registerrp-portal%2Caccount-portal%2Ccertificateprofile-portal%2Cdeleteresources-portal) to be used long-term for the windows code signing, as its the Microsoft recommended way and supported with official Github actions like https://github.com/azure/trusted-signing-action

sconklin committed:
Update documents and remove unused test workflows

sconklin committed:
Merge branch 'develop' into code-signing-pr-branch

sconklin commented:
> Overall I would have expected a lighter setup like Azure Trusted Signing (https://learn.microsoft.com/en-us/azure/trusted-signing/quickstart?tabs=registerrp-portal%2Caccount-portal%2Ccertificateprofile-portal%2Cdeleteresources-portal) to be used long-term for the windows code signing, as its the Microsoft recommended way and supported with official Github actions like https://github.com/azure/trusted-signing-action

Thanks for the review. Here are answers to several questions:

Seamly Systems, Inc is the owner of the certs, and the google HSM is on a Seamly corporate account. Sue and I have access to the key storage infrastructure.

Seamly Systems, Inc provides the certs and key storage as part of corporate sponsorship of the Seamly2D project, and pays the costs for those.

The reason for choosing Google HSM is that Seamly has no assets on Azure currently, and does have assets on both AWS and Google. The Google HSM setup was easier to set up, competitive for cost, and we already have ops accounts there so we picked it. It's been working really well all through testing, fwiw.

Also, I:

* Removed unused workflows left over from testing, leaving one test workflow
* Made the markdown files much smaller and removed all of the old planning documentation
* Changed some file names and updated documentation for clarity

After merging, this PR will require testing on the develop branch, which may result in multiple releases in a short time. I did the best I could in a test branch, but there are workflow dependencies on environments which need to be tested.

slspencer commented on .github/workflows/README.md:
Is there a reason why you removed the descriptive and instructive portions of this file?

slspencer commented on .github/workflows/ci.yml:
Should we use this opportunity to change the Version to our rolling weekly version based on date?
